Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,522 @@
+architectures and powerful abstractions. Like rules, generally, defining custom
+rules isn't required but can improve and consolidate functionality. For more
+information about aspects, see [Aspects the fan-fic of build
+rules](https://kchodorow.com/2017/01/10/aspects-the-fan-fic-of-build-rules/)
+
+### Toolchains
+
+In addition to configuring the build with rules, Bazel provides an additional
+primitive, the toolchain. Toolchains provide default compilers and arguments
+for those compilers. For the native c++ rules, this is relied upon to configure
+the many flags required for cross compilation.
+
+For most iOS projects, configuring the c++ toolchain isn't required. Defining a
+custom toolchain is useful to gives full control over compiler invocations or
+to customize the compiler being invoked.  Please see the [toolchain
+documentation](https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html)
+to learn more.
+
+### Generated Xcode projects
+
+In an Xcode world, folks checked in a project which contains an listing of
+files, build settings, and IDE state. When the project scales, the project file
+model breaks down. Auditing and code reviewing config changes in Xcode projects
+quickly becomes difficult. With Bazel, human readable `BUILD` files are the
+source of truth for build and Xcode configuration. Tools like
+[XCHammer](https://github.com/pinterest/xchammer), and
+[Tulsi](https://github.com/bazelbuild/tulsi) use an aspect to traverse the
+build graph and extract metadata required to generate a project.  These tools
+make it easier to manage the project and generate on demand - not needing to
+check it in.
+
+XCHammer provides a rule to [bazel build Xcode
+projects](ihttps://github.com/pinterest/xchammer#bazel-build-xcode-projects).
+Simply declare the rule with the project and Bazel build the target.
+```
+load("@xchammer_resources//:xcodeproject.bzl", "xcode_project")
+xcode_project(
+    name = "MyProject",
+    targets = [ "//ios-app:ios-app" ],
+    paths = [ "**" ],
+)
+```
+
+Like any other Bazel target, it's built from the command line
+```
+bazel build :MyProject
+```
+
+The rule definition for the `xcode_project` may be found in the [github
+repository](https://github.com/pinterest/xchammer/blob/master/BazelExtensions/xcodeproject.bzl).
+Simply put, the aspect traverses sources, and invokes the `xchammer` binary with
+a JSON file. Internally, XCHammer instantiates
+[XcodeGen](https://github.com/yonaskolb/XcodeGen) types and writes them out to
+disk with [xcodeproj](https://github.com/tuist/XcodeProj).
+
+Generating an Xcode project with XCHammer:
+```
+command line -> bazel -> rule -> xchammer -> xcodegen -> project.xcodeproj
+```
+
+Generators like XCHammer and Tulsi take care of integrating Bazel into the IDE.
+Bazel builds are invoked a shell script build phase from the IDE. Basically,
+Xcode shells out to Bazel to produce the application, and then Xcode picks up
```

Comment:
Does Xcode pick it up directly from derived data path? I guess it's worth mentioning that Tulsi manages that at least, otherwise one may think that it works out of the box with just a run script phase that builds the app.

Reply:
@BalestraPatrick This is a really good point, simply put, yes we currently rsync artifacts into the built products in derived data for xcode to pick it up.

For our internal simulator dev builds, there's a much more performant code path than what Xcode can do. We have a dynamic library architecture and integration where dyld3 loads dylibs residing in `bazel-out` via an rpath written into the binary. This is probably too in-depth and should be in a performance optimization segment along with the rules that allow this.